Add advertise_addrs.

i0rek opened the pull request:
Fixes #550.
This will make it possible to configure the advertised adresses for SerfLan, SerfWan and RPC. It will enable multiple consul clients on a single host which is very useful in a container environment.

This option might override `advertise_addr` and `advertise_addr_wan` depending on the configuration.

It will be configureable with `advertise_addrs`. Example:

```` json
{
  "advertise_addrs": {
    "serf_lan": "10.0.120.91:4424",
    "serf_wan": "201.20.10.61:4423",
    "rpc": "10.20.10.61:4424"
  }
}
```

ryanuber commented on command/agent/agent.go:
```diff
@@ -244,6 +244,24 @@ func (a *Agent) consulConfig() *consul.Config {
+			base.SerfWANConfig.MemberlistConfig.AdvertiseAddr = addr.IP.String()
+			base.SerfWANConfig.MemberlistConfig.AdvertisePort = addr.Port
+		}
+	}
```
Instead of printing a message for both of the above and allowing Consul to start even with a bad advertise address specified in its config, can we store the `*net.TCPAddr`s in the AdvertiseAddrs struct and parse the address while we are decoding? This would also make a config merge pretty easy for these fields. We do a similar thing with parsing time values.

ryanuber commented:
Hey @i0rek, I took a quick look through what you have so far and I think you're on the right track. We are missing a few things. See my comment above. We will also need to add the new config fields to the `Merge` method in a *Config, so that the new config options may be specified in any configuration file and not just the first one on the list.

i0rek commented:
Thank you very much for the suggestions @ryanuber! I will take care of them beginning next week.

i0rek commented:
@ryanuber I think I added everything you suggested. Still have to test it though.

ryanuber commented on command/agent/config.go:
```diff
@@ -638,6 +648,24 @@ func DecodeConfig(r io.Reader) (*Config, error) {
+		addr, err := net.ResolveTCPAddr("tcp", result.AdvertiseAddrs.SerfLanRaw)
+		if err != nil {
+			return nil, fmt.Errorf("AdvertiseAddrs.SerfLan is invalid: %v", err)
+		} else {
```
Minor nit-pick but let's drop these `else` statements, since we already return in the error path above.

i0rek replied:
I am sorry, I should've paid more attention to the other code.

ryanuber commented:
@i0rek this is looking good. Ping me when you feel it is complete and I'll give it another once-over!

i0rek commented:
In theory this could be necessary for CLI RPC, HTTP API and DNS Interface too. What do you think?

i0rek commented:
Ok, it works fine for our usecase. Which is letting consul run in two containers on one host, connecting to a remote server. This is the configuration I used: 

```
{ "advertise_addr": "10.111.111.112", "advertise_addrs": { "serf_lan": "10.111.111.112:9301" } }
```

I don't think RPC works yet, but I could add that afterwards.

```
docker1            10.111.111.112:9301    alive   client  0.5.2  2
docker2            10.111.111.112:8301    alive   client  0.5.2  2
```

i0rek commented:
@ryanuber I added RPC to the mix, because that was missing. I also added a little bit of documentation, not sure how you handle that. I was planning to rebase a last time after you looked at it so that you have a nice commit you can merge. I am not sure how you handle that either.

ryanuber commented on command/agent/config.go:
```diff
@@ -819,6 +855,15 @@ func MergeConfig(a, b *Config) *Config {
+	}
+	if b.AdvertiseAddrs.RPC != nil {
+		result.AdvertiseAddrs.RPC = b.AdvertiseAddrs.RPC
+	}
```
Looks like we are missing tests for this. Let's definitely make sure this is tested.

ryanuber commented on command/agent/config_test.go:
```diff
@@ -211,6 +211,36 @@ func TestDecodeConfig(t *testing.T) {
+	}
+	if config.AdvertiseAddrs.RPCRaw != "127.0.0.5:1234" {
+		t.Fatalf("bad: %#v", config)
+	}
```
In these tests we should also test that the `*net.TCPAddr` is populated correctly.

ryanuber commented:
@i0rek this looks good, just some minor notes about test coverage but overall this is really close! The documentation looks right. Also, no need to rebase this, since it can be cleanly applied against master in its current state.

i0rek commented:
Thanks for the review @ryanuber. I added the tests.

ryanuber commented:
Great! Will review again soon!

On June 13, 2015 3:40:35 PM PDT, Hans Hasselberg <notifications@github.com> wrote:
>Thanks for the review @ryanuber. I added the tests.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/hashicorp/consul/pull/1004#issuecomment-111757900

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

i0rek commented:
When I said rebasing, I meant squashing. I couldn't resist.

i0rek commented:
I added tests to agent_test to make sure the configured values are used.

i0rek commented:
@ryanuber any progress on this? Seems ready to merge to me. Thanks

ryanuber commented on website/source/docs/agent/options.html.markdown:
```diff
@@ -322,6 +322,16 @@ definitions support being updated during a reload.
   the [`-advertise` command-line flag](#_advertise).
 
+* <a name="advertise_addrs"></a><a href="#advertise_addrs">`advertise_addrs`</a> Allows to set
+  the advertised addresses for SerfLan, SerfLan and RPC together with the port. This gives
```
I think we meant for this to be `SerfLan, SerfWan, and RPC` ?

ryanuber commented:
@i0rek sorry for the delay! It's been busy. Left a super-minor comment but otherwise let's get this merged. Great work!

i0rek commented:
@ryanuber no worries. I changed the typo!

ryanuber commented:
LGTM, thanks again!

rotatingJazz commented:
BIG props!!! :+1:

i0rek commented:
:heart:
